### ISSUE TITLE
[optimization-intro.ipynb] problematic explanation of worse error than machine precision

### ISSUE THREAD

**Issue by dingran** [zackchase/mxnet-the-straight-dope#392]
In the current explanation below, it blames O(epsilon^2) for error larger than machine precision, but epsilon^2 is not going to be larger than epsilon (1e-8). The real reason is error magnification due to potentially large f'(x) and other things.

**Current explanation**

In fact, the precision of a solution to optimization can be worse than the machine precision. To demonstrate that, consider a function $f: \mathbb{R} \rightarrow \mathbb{R}$, its Taylor series exansion is

$$f(x + \epsilon) = f(x) + f'(x)\epsilon + \frac{f''(x)}{2}\epsilon^2 + \mathcal{O}(\epsilon^3)$$

where $\epsilon$ is small. Denote the global optimum solution as $x^*$ for minimizing $f(x)$. It usually holds that 

$$f'(x^*) = 0 \quad \text{and} \quad f''(x^*) \neq 0.$$

Thus, for a small value $\epsilon$, we have

$$f(x^* + \epsilon) \approx f(x^*) + \mathcal{O}(\epsilon^2),$$

where the coefficient term of $\mathcal{O}(\epsilon^2)$ is $f''(x)/2$. 
This means that a small change of order $\epsilon$ in the optimum solution $x^*$ 
will change the value of $f(x^*)$ in the order of $\epsilon^2$. 
In other words, if there is an error in the function value, 
the precision of solution value is constrained 
by the order of the square root of that error. 
For example, if the machine precision is $10^{-8}$, 
the precision of the solution value is only 
in the order of $10^{-4}$, 
which is much worse than the machine precision.
